File: apps/backend/app/schemas/notification.py
"""
通知系统Schema
"""

from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field
from enum import Enum


class NotificationTypeEnum(str, Enum):
    """通知类型枚举"""
    SYSTEM = "system"
    WORKFLOW = "workflow"
    REMINDER = "reminder"
    ANNOUNCEMENT = "announcement"
    MESSAGE = "message"


class NotificationChannelEnum(str, Enum):
    """通知渠道枚举"""
    INBOX = "inbox"
    EMAIL = "email"
    SMS = "sms"
    WEBHOOK = "webhook"


class NotificationPriorityEnum(str, Enum):
    """通知优先级枚举"""
    LOW = "low"
    NORMAL = "normal"
    HIGH = "high"
    URGENT = "urgent"


class NotificationStatusEnum(str, Enum):
    """通知状态枚举"""
    PENDING = "pending"
    SENT = "sent"
    READ = "read"
    FAILED = "failed"
    EXPIRED = "expired"


class NotificationResponse(BaseModel):
    """通知响应"""
    id: int
    type: NotificationTypeEnum
    title: str
    content: Optional[str] = None
    link: Optional[str] = None
    sender_id: Optional[int] = None
    receiver_id: int
    priority: NotificationPriorityEnum
    status: NotificationStatusEnum
    read_at: Optional[datetime] = None
    sent_at: Optional[datetime] = None
    expire_at: Optional[datetime] = None
    created_at: datetime

    class Config:
        from_attributes = True


class NotificationCreate(BaseModel):
    """创建通知"""
    receiver_id: int
    type: NotificationTypeEnum
    title: str
    content: Optional[str] = None
    link: Optional[str] = None
    priority: NotificationPriorityEnum = NotificationPriorityEnum.NORMAL
    expire_hours: Optional[int] = Field(None, description="过期小时数")


class NotificationBatchCreate(BaseModel):
    """批量创建通知"""
    receiver_ids: List[int]
    type: NotificationTypeEnum
    title: str
    content: Optional[str] = None
    link: Optional[str] = None
    priority: NotificationPriorityEnum = NotificationPriorityEnum.NORMAL
    expire_hours: Optional[int] = None


class NotificationTemplateResponse(BaseModel):
    """通知模板响应"""
    id: int
    code: str
    name: str
    type: NotificationTypeEnum
    title_template: str
    content_template: Optional[str] = None
    variables: Optional[List[str]] = None
    channels: Optional[List[str]] = None
    description: Optional[str] = None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class NotificationTemplateCreate(BaseModel):
    """创建通知模板"""
    code: str
    name: str
    type: NotificationTypeEnum
    title_template: str
    content_template: Optional[str] = None
    variables: Optional[List[str]] = None
    channels: Optional[List[str]] = None
    description: Optional[str] = None


class NotificationPreferenceResponse(BaseModel):
    """通知偏好设置响应"""
    id: int
    user_id: int
    notification_type: Optional[str] = None
    inbox_enabled: bool
    email_enabled: bool
    sms_enabled: bool
    quiet_hours_start: Optional[str] = None
    quiet_hours_end: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class NotificationPreferenceUpdate(BaseModel):
    """更新通知偏好设置"""
    notification_type: Optional[str] = None
    inbox_enabled: Optional[bool] = None
    email_enabled: Optional[bool] = None
    sms_enabled: Optional[bool] = None
    quiet_hours_start: Optional[str] = None
    quiet_hours_end: Optional[str] = None


class UnreadCountResponse(BaseModel):
    """未读消息数响应"""
    total: int
    by_type: dict

File: apps/frontend/src/types/notification.ts
/**
 * 通知系统类型定义
 * @description 通知相关的数据结构，用于前后端数据交互
 */

/**
 * 通知类型枚举
 */
export enum NotificationType {
  SYSTEM = "system",           // 系统通知
  WORKFLOW = "workflow",       // 工作流通知
  REMINDER = "reminder",       // 提醒通知
  ANNOUNCEMENT = "announcement", // 公告通知
  MESSAGE = "message",         // 私信
}

/**
 * 通知优先级枚举
 */
export enum NotificationPriority {
  LOW = "low",       // 低
  NORMAL = "normal", // 普通
  HIGH = "high",     // 高
  URGENT = "urgent", // 紧急
}

/**
 * 通知状态枚举
 */
export enum NotificationStatus {
  PENDING = "pending", // 待发送
  SENT = "sent",       // 已发送
  READ = "read",       // 已读
  FAILED = "failed",   // 发送失败
  EXPIRED = "expired", // 已过期
}

/**
 * 通知响应数据结构
 */
export interface Notification {
  id: number;
  type: NotificationType;
  title: string;
  content?: string;
  link?: string;
  sender_id?: number;
  receiver_id: number;
  priority: NotificationPriority;
  status: NotificationStatus;
  read_at?: string;
  sent_at?: string;
  expire_at?: string;
  created_at: string;
}

/**
 * 创建通知请求
 */
export interface CreateNotification {
  receiver_id: number;
  type: NotificationType;
  title: string;
  content?: string;
  link?: string;
  priority?: NotificationPriority;
  expire_hours?: number;
}

/**
 * 未读消息数响应
 */
export interface UnreadCount {
  total: number;
  by_type: Record<string, number>;
}

/**
 * 通知列表响应
 */
export interface NotificationListResponse {
  items: Notification[];
  total: number;
  page: number;
  page_size: number;
}

/**
 * 通知偏好设置
 */
export interface NotificationPreference {
  id: number;
  user_id: number;
  notification_type?: string;
  inbox_enabled: boolean;
  email_enabled: boolean;
  sms_enabled: boolean;
  quiet_hours_start?: string;
  quiet_hours_end?: string;
  created_at: string;
  updated_at: string;
}

/**
 * 通知偏好设置更新请求
 */
export interface UpdateNotificationPreference {
  notification_type?: string;
  inbox_enabled?: boolean;
  email_enabled?: boolean;
  sms_enabled?: boolean;
  quiet_hours_start?: string;
  quiet_hours_end?: string;
}
